akka-stream/src/main/scala/akka/stream/stage/Stage.scala

```diff
@@ -306,6 +310,125 @@ abstract class AsyncStage[In, Out, Ext]
+    queued.reverse.foreach {
+      case QueuedSchedule(timerKey, interval)  ⇒ schedulePeriodically(timerKey, interval)
+      case QueuedScheduleOnce(timerKey, delay) ⇒ scheduleOnce(timerKey, delay)
+      case QueuedCancelTimer(timerKey)         ⇒ cancelTimer(timerKey)
+    }
+    queued = Nil
+  }
+
+  override def postStop(): Unit = {
+    super.postStop()
+    timers.foreach { case (_, Timer(_, task)) ⇒ task.cancel() }
+    timers.clear()
+  }
+
+  override def onAsyncInput(scheduled: Scheduled, ctx: AsyncContext[Out, Scheduled]): Directive = {
+    val Id = scheduled.timerId
+    timers.get(scheduled.timerKey) match {
+      case Some(Timer(Id, _)) ⇒
+        if (!scheduled.repeating) timers -= scheduled.timerKey
+        onTimer(scheduled.timerKey, ctx)
+      case _ ⇒ ctx.ignore()
+    }
+  }
+
+  /**
+   * Schedule timer to call [[#onTimer]] periodically with the given interval.
+   * Any existing timer with the same key will automatically be canceled before
+   * adding the new timer.
+   */
+  final def schedulePeriodically(timerKey: Any, interval: FiniteDuration): Unit =
+    system match {
+      case Some(ctx) ⇒
+        cancelTimer(timerKey)
+        val id = timerIdGen.next()
+        val task = ctx.scheduler.schedule(interval, interval) {
+          callback.invoke(Scheduled(timerKey, id, repeating = true))
+        }(ctx.dispatcher)
+        timers(timerKey) = Timer(id, task)
+      case None ⇒
+        queued = QueuedSchedule(timerKey, interval) :: queued
+    }
+
+  /**
+   * Schedule timer to call [[#onTimer]] after given delay.
+   * Any existing timer with the same key will automatically be canceled before
+   * adding the new timer.
+   */
+  final def scheduleOnce(timerKey: Any, delay: FiniteDuration): Unit =
+    system match {
+      case Some(ctx) ⇒
+        cancelTimer(timerKey)
+        val id = timerIdGen.next()
+        val task = ctx.scheduler.scheduleOnce(delay) {
+          callback.invoke(Scheduled(timerKey, id, repeating = false))
+        }(ctx.dispatcher)
+        timers(timerKey) = Timer(id, task)
+      case None ⇒
+        queued = QueuedScheduleOnce(timerKey, delay) :: queued
+    }
+
+  /**
+   * Cancel timer, ensuring that the [[#onTimer]] is not subsequently called.
+   * @param timerKey key of the timer to cancel
+   */
+  final def cancelTimer(timerKey: Any): Unit =
+    timers.get(timerKey).foreach { t ⇒
+      t.task.cancel()
+      timers -= timerKey
+    }
+
+  /**
+   * Inquire whether the timer is still active. Returns true unless the
+   * timer does not exist, has previously been canceled or if it was a
+   * single-shot timer that was already triggered.
+   */
+  final def isTimerActive(timerKey: Any): Boolean = timers contains timerKey
+
+  /**
+   * Will be called when the scheduled timer is triggered.
+   * @param timerKey key of the scheduled timer
+   */
+  def onTimer(timerKey: Any, ctx: AsyncContext[Out, Scheduled]): Directive
```

Comment:
`Scheduled` is `private[stream]` but it is exposed in the public method, is that alright?